### PR TITLE
fix(exceptions): avoid swallowing results

### LINKED ISSUE #1018: Can't silence reports
**Describe the bug**

I just installed 5.2.0 with the new silencing feature introduced by @nulltoken. I'm trying to silence the errors I am getting for my openapi.yaml described here: #915

To start with broad example, I added
```
except:
  "openapi.yaml#":
    - oas3-valid-schema-example
    - oas3-valid-content-schema-example
```
Unfortunately, it still shows the errors. I also tried some more paths, but except for an error that the source is missing for path `#` nothing changed.

What I really want to achieve:
```
except:
  "openapi.yaml#/components/schemas/process_graph":
    - oas3-valid-schema-example
  "openapi.yaml#/paths/\/processes/get":
    - oas3-valid-content-schema-example
  "openapi.yaml#/paths/\/process_graphs/get":
    - oas3-valid-content-schema-example
  "openapi.yaml#/paths/\/service_types/get":
    - oas3-valid-content-schema-example
```
Although for this example I'm not sure whether the forward slashes are correctly escaped.

I'm using JSON Reference here because the example in the docs look like JSON Reference although the doc says JSON Path. I'm confused, but tried both...

**To Reproduce**

See #915 and add the code given above.

**Expected behavior**

Errors should be silent. Either I don't understand the new feature/docs and that needs improvement or there's a bug in the code.

**Environment (remove any that are not applicable):**
 - Library version: 5.2.0
 - OS: Windows 10

### PR DESCRIPTION
**Pull request by nulltoken**
Fix #1018

**Checklist**

- [x] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

/cc @m-mohr

### CODE REVIEW

**Comment by nulltoken** [comment]
@m-mohr And here's a fix proposal.

You can give it an early try (and play a little with it to ensure it now works a little bit better) by checking this branch out, maybe running `yarn` to ensure your deps are up to date then:

```
$  yarn cli lint ./repro/openapi.yaml -r ./repro/ruleset.yaml
```

where the `ruleset` would be expressed as described in https://github.com/stoplightio/spectral/issues/1018#issuecomment-601011695

_Note:_ passing `-f json` to the cli usually makes it easier to figure out the paths one want to `except`

**Comment by m-mohr** [comment]
Thanks, @nulltoken!

> _Note:_ passing `-f json` to the cli usually makes it easier to figure out the paths one want to `except`

Isn't this something that should be added to the docs? Also, I think it would be useful to say that the path needs to be "precise". It seems you can't just point to any parent, right?

File: docs/getting-started/rulesets.md [hunk 1]
```diff
@@ -256,7 +256,7 @@ the rule entirely. This may happen, for instance, when working with legacy APIs.
 
 The ruleset can be extended for that purpose through the optional `except` property.
 
-`except` describes a map of locations (expressed as Json paths) and rules that should be ignored.
+`except` describes a map of locations (expressed paths and Json pointers) and rules that should be ignored.
```
Comment (philsturgeon) [inline docs/getting-started/rulesets.md]:
```suggestion
`except` describes a map of locations (expressed paths and [JSON Pointers]((https://tools.ietf.org/html/rfc6901)) and rules that should be ignored.
```

File: docs/getting-started/rulesets.md [hunk 2]
```diff
@@ -270,6 +270,9 @@ except:
     - info-contact
     - info-description
 ```
+_Note:_ As per the [RFC 6901](https://tools.ietf.org/html/rfc6901#section-3), special characters
```
Comment (philsturgeon) [inline docs/getting-started/rulesets.md]:
```suggestion

<!-- theme: info -->
> As per the [RFC 6901](https://tools.ietf.org/html/rfc6901#section-3), special characters
```

File: docs/getting-started/rulesets.md [hunk 3]
```diff
@@ -270,6 +270,9 @@ except:
...
+`~` and `/` have to be escaped to `~0` and `~1` respectively. For instance, the location of the `get` method from a
```
Comment (philsturgeon) [inline docs/getting-started/rulesets.md]:
```suggestion
> `~` and `/` have to be escaped to `~0` and `~1` respectively. For instance, the location of the `get` method from a
```

File: docs/getting-started/rulesets.md [hunk 4]
```diff
@@ -270,6 +270,9 @@ except:
...
+`/todos` path in an openapi document `/root/here.yaml` would be expressed as `/root/here.yaml#/paths/~1todos/get`.
```
Comment (philsturgeon) [inline docs/getting-started/rulesets.md]:
```suggestion
> `/todos` path in an openapi document `/root/here.yaml` would be expressed as `/root/here.yaml#/paths/~1todos/get`.
```

sorry coudnt make multi line suggestions work today

File: docs/getting-started/rulesets.md (hunk 2 above)
Comment (nulltoken) [reply docs/getting-started/rulesets.md]: Fixed

File: docs/getting-started/rulesets.md (hunk 3 above)
Comment (nulltoken) [reply docs/getting-started/rulesets.md]: Fixed

File: docs/getting-started/rulesets.md (hunk 4 above)
Comment (nulltoken) [reply docs/getting-started/rulesets.md]: Fixed

File: docs/getting-started/rulesets.md [hunk 5]
```diff
@@ -271,6 +272,13 @@ except:
     - info-description
 ```
 
+<!-- theme: info -->
+> As per the [RFC 6901](https://tools.ietf.org/html/rfc6901#section-3), special characters
+> `~` and `/` have to be escaped to `~0` and `~1` respectively. For instance, the location of the `get` method from a
+> `/todos` path in an openapi document `/root/here.yaml` would be expressed as `/root/here.yaml#/paths/~1todos/get`.
+> 
+> _Hint:_ Running Spectral cli with the `--format json` parameter is pretty useful to find out the path segments of each result.
```
Comment (nulltoken) [inline docs/getting-started/rulesets.md]: As pointed out by @m-mohr , added this small hint.

@philsturgeon Does the wording fit you?

File: src/linter.ts [hunk 6]
```diff
@@ -19,13 +19,18 @@ const isAKnownException = (violation: IRuleResult, locations: IExceptionLocation
       continue;
     }
 
-    for (let i = 0; i < violation.path.length; i++) {
+    let arePathEquals: boolean = true;
```
Comment (nulltoken) [inline src/linter.ts]: Dammit. Those ten lines of code were wrong is so many ways...

Thanks @m-mohr for all the help and feedback!

**Comment by nulltoken** [comment]
> Let's get some test harness tests on this new feature, specifically using the examples from #1018 so they're not contrived.

Yep. Will add that tomorrow.

**Comment by nulltoken** [comment]
@m-mohr New version with enhanced fix is ready to test, Sir!

Running `yarn cli lint ./repro/openapi.yaml -r ./repro/ruleset.yaml`

with the following ruleset 

```
extends: spectral:oas

except:
  "openapi.yaml#/components/schemas/process_graph/example":
    - oas3-valid-schema-example
  "openapi.yaml#/paths/~1process_graphs/get/responses/200/content/application~1json/schema/example":
    - oas3-valid-content-schema-example
  "openapi.yaml#/paths/~1service_types/get/responses/200/content/application~1json/schema/example":
    - oas3-valid-content-schema-example
  "openapi.yaml#/paths/~1processes/get/responses/200/content/application~1json/schema/example":
    - oas3-valid-content-schema-example
```

now returns zero result.

**Comment by m-mohr** [comment]
Thanks, @nulltoken! I can confirm that it works as expected. :-) Looking forward to 5.2.1.

**Comment by nulltoken** [comment]
> > Let's get some test harness tests on this new feature, specifically using the examples from #1018 so they're not contrived.
> 
> Yep. Will add that tomorrow.

@philsturgeon Done. The initial repro from #915 was too complicated to trim down. I rewrote something similar and used it to create to harness scenarii (a reference one and a silenced one).

File: docs/getting-started/rulesets.md (hunk 5 above)
Comment (philsturgeon) [reply docs/getting-started/rulesets.md]: Sure

File: src/linter.ts [hunk 7]
```diff
@@ -19,13 +19,18 @@ const isAKnownException = (violation: IRuleResult, locations: IExceptionLocation
...
+
+    for (let i = 0; arePathEquals && i < violation.path.length; i++) {
       if (location.path[i] !== violation.path[i]) {
+        arePathEquals = false;
```
Comment (P0lip) [inline src/linter.ts]:
```suggestion
    let arePathEquals: boolean = true;

    for (let i = 0; arePathEquals && i < violation.path.length; i++) {
      if (location.path[i] !== violation.path[i]) {
        return false;
```

File: src/linter.ts [hunk 8]
```diff
@@ -19,13 +19,18 @@ const isAKnownException = (violation: IRuleResult, locations: IExceptionLocation
...
       }
     }
 
-    return true;
+    if (arePathEquals) {
+      return true;
+    }
   }
 
   return false;
```
Comment (P0lip) [inline src/linter.ts]:
```suggestion
  return true;
```

File: src/linter.ts [hunk 9]
```diff
@@ -19,13 +19,18 @@ const isAKnownException = (violation: IRuleResult, locations: IExceptionLocation
...
       if (location.path[i] !== violation.path[i]) {
+        arePathEquals = false;
         continue;
       }
     }
 
-    return true;
+    if (arePathEquals) {
+      return true;
+    }
```
Comment (P0lip) [inline src/linter.ts]:
```suggestion
```

File: src/linter.ts (hunk 7 above)
Comment (nulltoken) [reply src/linter.ts]: > One suggestion that shouldn't result in a behavior change, I believe.

@P0lip I'm afraid it does. :-/ if we return `false` here, we won't process the potentially remaining locations. Let me see if I can refactor this a bit to get rid of those nested loops and make the intent clearer.

Comment (nulltoken) [reply src/linter.ts]: >Let me see if I can refactor this a bit to get rid of those nested loops and make the intent clearer.

@P0lip Done. Do you like it better?

Comment (P0lip) [reply src/linter.ts]: @nulltoken it's better!

File: docs/getting-started/rulesets.md [hunk 10]
```diff
@@ -256,7 +256,8 @@ the rule entirely. This may happen, for instance, when working with legacy APIs.
 
 The ruleset can be extended for that purpose through the optional `except` property.
 
-`except` describes a map of locations (expressed as Json paths) and rules that should be ignored.
+`except` describes a map of locations (expressed as paths and [JSON Pointers]((https://tools.ietf.org/html/rfc6901))
```
Comment (P0lip) [inline docs/getting-started/rulesets.md]:
```suggestion
`except` describes a map of locations (expressed as paths and [JSON Pointers](https://tools.ietf.org/html/rfc6901))
```

**Pull request merged**